File: evidence/v1_5/ai_benchmark/subjects/primary/r2/rle/cpp/main.cpp
#include <cstdint>
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>

static int hexVal(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    return c - 'a' + 10;
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string capLine, hexLine;
    std::getline(std::cin, capLine);
    std::getline(std::cin, hexLine);

    uint64_t cap = 0;
    for (char c : capLine) {
        if (c < '0' || c > '9') continue;
        cap = cap * 10 + (uint64_t)(c - '0');
    }

    std::vector<uint8_t> in;
    in.reserve(hexLine.size() / 2);
    for (size_t i = 0; i + 1 < hexLine.size(); i += 2) {
        uint8_t b = (uint8_t)((hexVal(hexLine[i]) << 4) | hexVal(hexLine[i + 1]));
        in.push_back(b);
    }

    size_t n = in.size();
    size_t i = 0;
    uint64_t outLen = 0;
    uint64_t hash = 14695981039346656037ULL;

    while (i < n) {
        size_t offset = i;
        uint8_t c = in[i];
        if (c < 128) {
            size_t len = (size_t)c + 1;
            if (i + 1 + len > n) {
                std::cout << "err truncated " << offset << "\n";
                return 0;
            }
            if (outLen + len > cap) {
                std::cout << "err full " << offset << "\n";
                return 0;
            }
            for (size_t k = 0; k < len; k++) {
                uint8_t b = in[i + 1 + k];
                hash = (hash ^ b) * 1099511628211ULL;
            }
            outLen += len;
            i += 1 + len;
        } else {
            size_t count = (size_t)c - 126;
            if (i + 1 >= n) {
                std::cout << "err truncated " << offset << "\n";
                return 0;
            }
            if (outLen + count > cap) {
                std::cout << "err full " << offset << "\n";
                return 0;
            }
            uint8_t b = in[i + 1];
            for (size_t k = 0; k < count; k++) {
                hash = (hash ^ b) * 1099511628211ULL;
            }
            outLen += count;
            i += 2;
        }
    }

    char buf[17];
    std::snprintf(buf, sizeof(buf), "%016llx", (unsigned long long)hash);
    std::cout << "ok " << outLen << " " << buf << "\n";
    return 0;
}

File: evidence/v1_5/ai_benchmark/subjects/primary/r2/rle/rust/main.rs
use std::io::{self, Read};

fn read_input() -> String {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    s
}

fn hex_val(b: u8) -> u8 {
    match b {
        b'0'..=b'9' => b - b'0',
        b'a'..=b'f' => b - b'a' + 10,
        _ => 0,
    }
}

fn main() {
    let input = read_input();
    let mut lines = input.split('\n');
    let cap_line = lines.next().unwrap_or("").trim_end_matches('\r').trim();
    let hex_line = lines.next().unwrap_or("").trim_end_matches('\r').trim();

    let cap: usize = cap_line.parse().unwrap();

    let hex_bytes = hex_line.as_bytes();
    let n = hex_bytes.len() / 2;
    let mut data: Vec<u8> = Vec::with_capacity(n);
    for i in 0..n {
        let hi = hex_val(hex_bytes[2 * i]);
        let lo = hex_val(hex_bytes[2 * i + 1]);
        data.push((hi << 4) | lo);
    }

    let mut out_len: usize = 0;
    let mut hash: u64 = 14695981039346656037;
    let mut i: usize = 0;

    while i < data.len() {
        let offset = i;
        let c = data[i];
        if c < 128 {
            let count = (c as usize) + 1;
            if i + 1 + count > data.len() {
                println!("err truncated {}", offset);
                return;
            }
            if out_len + count > cap {
                println!("err full {}", offset);
                return;
            }
            for k in 0..count {
                let b = data[i + 1 + k];
                hash = (hash ^ (b as u64)).wrapping_mul(1099511628211);
            }
            out_len += count;
            i += 1 + count;
        } else {
            let count = (c as usize) - 126;
            if i + 1 >= data.len() {
                println!("err truncated {}", offset);
                return;
            }
            if out_len + count > cap {
                println!("err full {}", offset);
                return;
            }
            let b = data[i + 1];
            for _ in 0..count {
                hash = (hash ^ (b as u64)).wrapping_mul(1099511628211);
            }
            out_len += count;
            i += 2;
        }
    }

    println!("ok {} {:016x}", out_len, hash);
}
